File: modules/vpc/src/main/scala/io/github/mvillafuertem/backend/Backend.scala
package io.github.mvillafuertem.backend

import com.hashicorp.cdktf.TerraformOutput
import io.github.mvillafuertem.aws.dynamodb_table.{ DynamodbTable, DynamodbTableAttribute }
import io.github.mvillafuertem.aws.s3_bucket._
import io.github.mvillafuertem.aws.s3_bucket_server_side_encryption_configuration.{
  S3BucketServerSideEncryptionConfigurationA,
  S3BucketServerSideEncryptionConfigurationRuleA,
  S3BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDefaultA
}
import io.github.mvillafuertem.aws.s3_bucket_versioning.{ S3BucketVersioningA, S3BucketVersioningVersioningConfiguration }
import io.github.mvillafuertem.backend.Backend.BackendConfiguration
import software.constructs.Construct

import scala.jdk.CollectionConverters._

final case class Backend(configuration: BackendConfiguration)(implicit scope: Construct, id: String) extends Construct(scope, id) {
  self: Construct =>

  private val s3Bucket: S3Bucket = S3Bucket.Builder
    .create(self, "s3")
    .bucket(configuration.bucket)
    .forceDestroy(true)
    .tags(configuration.commonTags.asJava)
    .build()

  S3BucketServerSideEncryptionConfigurationA.Builder
    .create(self, "s3_server_side_encryption")
    .bucket(s3Bucket.getBucket)
    .rule(
      Seq(
        S3BucketServerSideEncryptionConfigurationRuleA
          .builder()
          .applyServerSideEncryptionByDefault(
            S3BucketServerSideEncryptionConfigurationRuleApplyServerSideEncryptionByDefaultA
              .builder()
              .sseAlgorithm("AES256")
              .build()
          )
          .build()
      ).asJava
    )
    .build()

  S3BucketVersioningA.Builder
    .create(self, "s3_versioning")
    .bucket(s3Bucket.getBucket)
    .versioningConfiguration(
      S3BucketVersioningVersioningConfiguration
        .builder()
        .status("Enabled")
        .build()
    )
    .build()

  private val dynamodbTable: DynamodbTable = DynamodbTable.Builder
    .create(self, "dynamodb")
    .name(configuration.dynamodbTable)
    .billingMode("PAY_PER_REQUEST")
    .hashKey("LockID")
    .attribute(List(DynamodbTableAttribute.builder().name("LockID").`type`("S").build()).asJava)
    .tags(configuration.commonTags.asJava)
    .build()

  private val _: TerraformOutput = TerraformOutput.Builder
    .create(self, "tfstate_output")
    .value(s"${s3Bucket.getArn}")
    .build()

}

object Backend {

  final case class BackendConfiguration(
    bucket: String,
    dynamodbTable: String,
    commonTags: Map[String, String]
  )

}

File: libs/cdktf-ts/src/lib/infrastructure/backend.ts
import {Construct} from "constructs";
import {S3Bucket} from "../../../io.github.mvillafuertem/providers/aws/s3-bucket";
import {
    S3BucketServerSideEncryptionConfigurationA
} from "../../../io.github.mvillafuertem/providers/aws/s3-bucket-server-side-encryption-configuration";
import {S3BucketVersioningA} from "../../../io.github.mvillafuertem/providers/aws/s3-bucket-versioning";
import {DynamodbTable} from "../../../io.github.mvillafuertem/providers/aws/dynamodb-table";
import {TerraformOutput} from "cdktf";

interface BackendConfiguration {
    bucket: string,
    dynamodbTable: string,
    commonTags: { [key: string]: string },
}

export class BackendStack extends Construct {
    constructor(scope: Construct, id: string, backendConfiguration: BackendConfiguration) {
        super(scope, id);

        const {bucket, dynamodbTable} = backendConfiguration;

        const s3Bucket: S3Bucket = new S3Bucket(this, "s3", {
            bucket: bucket,
            forceDestroy: true,
            tags: {
                owner: "cdktf",
            }
        });

        new S3BucketServerSideEncryptionConfigurationA(this, "s3_server_side_encryption", {
            bucket: s3Bucket.bucket,
            rule: [{
                applyServerSideEncryptionByDefault: {
                    sseAlgorithm: "AES256"
                }
            }]
        })

        new S3BucketVersioningA(this, "s3_versioning", {
            bucket: s3Bucket.bucket,
            versioningConfiguration: {
                status: "Enabled"
            }
        })

        new DynamodbTable(this, "dynamodb", {
            name: dynamodbTable,
            billingMode: "PAY_PER_REQUEST",
            hashKey: "LockID",
            attribute: [{name: "LockID", type: "S"}],
            tags: {}
        })

        new TerraformOutput(this, "tfstate_output", {
            value: s3Bucket.arn
        })

    }

}